pan-domain-auth-verification/src/main/scala/com/gu/pandomainauth/service/CryptoConf.scala

```diff
@@ -42,4 +74,33 @@ object CryptoConf {
     def privateKeyFor(base64Key: String): SettingsResult[PrivateKey] =
       keyFor(base64Key, new PKCS8EncodedKeySpec(_), _.generatePrivate)
   }
+
+  object Change {
```

Comment:
I found it quite difficult to parse the logic in this function. I feel as though it could be easier to read if it was made a little less concise, e.g. something like:

```
      if (newConf == oldConf) {
        None
      } else {
        val maybeActiveKey = if (newConf.activeKeyPair == oldConf.activeKeyPair) {
          None
        } else {
          Some(
            ActiveKey(
              toleratingOldKey = newConf.alsoAccepted.contains(oldConf.activeKeyPair.publicKey),
              newKeyAlreadyAccepted = oldConf.alsoAccepted.contains(newConf.activeKeyPair.publicKey)
            )
          )
        }
        Some(Change(
          activeKey = maybeActiveKey,
          SeqDiff.compare(oldConf.alsoAccepted, newConf.alsoAccepted)
        ))
      }
```

However, even once I'd parsed the logic I found the `Change` and `ActiveKey` abstractions a bit difficult to understand. Possibly it could be made a little easier by adjusting some naming, but if I understand correctly these classes are only used for generating logging messages, so I wonder if it'd be worth considering something a bit less elegant but perhaps more digestible, e.g.

```
val oldAndNewKeysAreTheSame = ???
val oldKeyIsDifferentButStillTolerated = ???
val newKeyIsAlreadyAccepted = ???

val (added, removed) = SeqDiff.compare(???)

val summary = ???
```

Maybe my suggestion above is a bit _too_ plodding, but I wonder if there's a happy middle ground available?

Reply:
thanks very much for the detailed explanation!

```scala
activeKey = Option.when(newConf.activeKeyPair != oldConf.activeKeyPair)(ActiveKey(
      toleratingOldKey = newConf.acceptsActiveKeyFrom(oldConf),
      newKeyAlreadyAccepted = oldConf.acceptsActiveKeyFrom(newConf)
)),
```

this alternative feels much more readable and explicit to me!